Show a warning when engine is in swarm mode

aanand opened the pull request:
Prompted by issues such as https://github.com/docker/docker/issues/23885.

```console
$ docker swarm init

$ docker-compose up
WARNING: The Docker Engine you're using is part of a swarm.
Compose does not make use of swarm mode to deploy services - all containers will be scheduled on the current node.
If you want to deploy your application across the swarm, use the bundle feature of the Docker experimental build. More info:

https://github.com/docker/docker/tree/master/experimental

...[rest of output]...
```

A few things:

- Thoughts on the wording? It feels a bit awkward to me.
- Should "swarm" be capitalised?
- There's no stable URL for the [stacks and bundles](https://github.com/docker/docker/blob/master/experimental/docker-stacks-and-bundles.md) guide - that file will presumably be moved into the docs subdirectory as soon as it exits experimental status. What can we do about that?
- Are the unit tests overkill?

ping @dnephin @aluzzardi @sfsmithcha @mgoelzer @samalba @londoncalling

aluzzardi commented:
Feature LGTM

Wording SGTM - @mgoelzer?

sfsmithcha commented:
In general wording LGTM,

minor edits:
```
$ docker-compose up
WARNING: The Docker Engine you're using is running in swarm mode.

Compose does not use swarm mode to deploy services to multiple nodes in a swarm.  All containers will be scheduled on the current node.

To deploy your application across the swarm, use the bundle feature of the Docker experimental build. 

More info:
https://github.com/docker/docker/tree/master/experimental
```

aanand commented:
Perhaps we should set up a permanent URL

@sfsmithcha Wording changes look good. Updated.

aanand commented:
Whoops, my finger slipped.

Perhaps we should set up a permanent URL for the bundle docs which redirects to the experimental doc for the time being, but which we can switch over to the new location once bundles are part of the main release.

We should also probably not say that it's experimental in the error message, because eventually it won't be, and then users of older versions of Compose will be misinformed.

sfsmithcha commented:
@aanand one option is that we could potentially move the experimental doc into the Compose docs with a banner at the top of the page that indicates the feature is experimental. Then it could live at its intended destination without need for redirect.

cc: @londoncalling @thaJeztah

thaJeztah commented:
@aanand we can use the "tag" instead of master, e.g. https://github.com/docker/docker/tree/v1.12.0-rc2/experimental

But we should decide if we want the experimental features in the regular docs (with a banner), or on GitHub only; currently, the "plugin" docs are in the actual documentation, but with a banner, so we should be consistent there

dnephin commented on compose/cli/command.py:
```diff
@@ -80,6 +80,7 @@ def get_client(environment, verbose=False, version=None, tls_config=None, host=N
         version=version, tls_config=tls_config, host=host,
         environment=environment, tls_version=get_tls_version(environment)
     )
+    warn_for_swarm_mode(client)
```
I don't think we want a warning on every command, just on `up`.

If someone has gone ahead and run `up` and knows what to expect, we can let them proceed without warning them ever time.

Maybe `scale` as well as `up` ? But i think at most those two.

aanand replied:
OK, that makes sense to me.

> If someone has gone ahead and run up and knows what to expect, we can let them proceed without warning them ever time.

We could check for existing containers, and only print the warning if there aren't any?

dnephin replied:
Sorry, I just meant that a warning on `up` would be enough. Reading it again I realized it wasn't clear.

I think we still need to warn even if there are containers. They might have started the containers before enabling swarm mode. Then if they run `up` again (thinking it will move the containers onto other nodes or something), the warning would be important.

aanand commented:
Updated so it only warns on `up`. I think the only remaining blocker is deciding what to do about the doc URL.

@thaJeztah: Pinning to the tag is good - better than linking to master, anyhow - but if we do that, it'll be impossible to update the document that we link them to.

I think you're right that we should be consistent with how we document experimental features. I think this issue presents a case for documenting them in the regular docs with a banner - we then have a page with a stable URL but which can still evolve (e.g. to remove the banner or redirect elsewhere).

thaJeztah commented:
@aanand yes, it's a tricky one, and not 100% sure what the best solution is. At least pinning the version makes sure users don't end up with a 404

aanand commented:
How about:

1. Add [docker-stacks-and-bundles.md](https://github.com/docker/docker/blob/v1.12.0-rc3/experimental/docker-stacks-and-bundles.md) to the Compose docs (under, say, `/compose/bundles`), with a banner declaring it to be experimental, and linking back to the same file in docker/docker.

2. Link to `http://docs.docker.com/compose/bundles` in the warning text.

3. As and when `docker deploy` exits experimental status, and a comprehensive bundle doc page is added to the Engine docs proper, revise the Compose page accordingly and link to the Engine page. Or just redirect straight there - we can decide later.

aanand commented:
Bump. If there are no objections, I'm going to do what I described above.

thaJeztah commented:
Oh, sorry, looks like I missed that comment.

Adding to the docs sounds good to me; note that you can mark a specific page as "experimental" by setting some extra metadata in the file, see https://github.com/docker/docker/blame/master/docs/reference/commandline/stack_deploy.md#L6-L12 (`advisory = "experimental"`); we also added `(experimental)` to the headings to prevent confusion when reading those docs on GitHub

thaJeztah commented:
We may want to keep a reference to the experimental features inside the `/experimental` directory (for people that are looking for which features are currently experimental)

aanand commented:
OK, I've copied in the bundle docs with an `advisory = "experimental"` annotation and updated the URL in the error message.

Naturally, the URL won't work until this PR is merged and the commit is published to the docs site.

aanand commented:
@londoncalling @SvenDowideit Any idea why the docs build is failing?

https://jenkins.dockerproject.org/job/docs-compose-pr/1354/

thaJeztah commented:
looks like docs failing is due to this Hugo bug; https://github.com/spf13/hugo/issues/1754

thaJeztah commented:
retest this please

aanand commented:
Thanks @thaJeztah - build's green now.

thaJeztah commented:
LGTM

SvenDowideit commented:
personally, I'd much rather all experimental docs were also on docs.docker.com, they become searchable, and are archived together with the rest of what applies to a specific release :)

SvenDowideit commented on docs/bundles.md:
```diff
@@ -0,0 +1,199 @@
+        with full content hash to fully specify the deployment artifact for the
+        service. Example:
+        <code>postgres@sha256:f76245b04ddbcebab5bb6c28e76947f49222c99fec4aadb0bb
+        1c24821a 9e83ef</code>
```
what are these extra hashs? in my dsb I only get a single hash like `"Image": "redis@sha256:b50f15d427aea5b579f9bf972ab82ff8c1c47bffc0481b225c6a714095a9ec34",`

SvenDowideit commented on docs/bundles.md:
```diff
@@ -0,0 +1,199 @@
+    <dd>
+        Networks that the service containers should be connected to. An entity
+        deploying a bundle should create networks as needed.
+    </dd>
```
is there a volumes section? - if not, its work mentioning so readers don't wonder if its an accidental omission.

aanand commented:
@SvenDowideit Addressed your comments. Once this PR is merged, I'll open another against docker/docker to bring `experimental/docker-stacks-and-bundles.md` in line.

aanand commented:
ping @shin- @dnephin. Further docs changes aren't urgent, but the code needs to get in before release.

dnephin commented on docs/bundles.md:
````diff
@@ -0,0 +1,200 @@
+Run 'docker stack COMMAND --help' for more information on a command.
+```
+
+## Bundle file format
````
Are we sure we want to document this?

It is likely to change significantly, and I don't think we expect anyone to be crafting this directly. 

I guess it doesn't hurt. We might want to document that we don't expect anyone to create a file directly.

dnephin commented:
LGTM
